### PR TITLE
[OpenMPIRBuilder] Add ThreadLimit and NumTeams clauses to teams construct

### PR DESCRIPTION
**Pull request by shraiysh**
This patch adds support for `thread_limit` and bounds on `num_teams` clause for the teams construct in OpenMP.

Added testcases for the same.

### CODE REVIEW

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h [hunk 1]
```diff
@@ -1917,8 +1917,13 @@ class OpenMPIRBuilder {
   ///
   /// \param Loc The location where the teams construct was encountered.
   /// \param BodyGenCB Callback that will generate the region code.
+  /// \param NumTeamsUpper Upper bound on the number of teams.
+  /// \param ThreadLimit on the number of threads that may participate in a
+  ///        contention group created by each team.
   InsertPointTy createTeams(const LocationDescription &Loc,
-                            BodyGenCallbackTy BodyGenCB);
+                            BodyGenCallbackTy BodyGenCB,
+                            Value *NumTeamsUpper = nullptr,
```
Comment (kiranchandramohan) [inline llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1925]: What about the LowerBound?

File: llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp [hunk 2]
```diff
@@ -4082,6 +4082,163 @@ TEST_F(OpenMPIRBuilderTest, CreateTeams) {
...
+                       GlobalValue::ExternalLinkage, "fakeFunction", M.get());
+
+  auto BodyGenCB = [&](InsertPointTy AllocaIP, InsertPointTy CodeGenIP) {
+    Builder.restoreIP(CodeGenIP);
+    Builder.CreateCall(FakeFunction, {});
+  };
+
+  OpenMPIRBuilder::LocationDescription Loc({Builder.saveIP(), DL});
+  Builder.restoreIP(
+      OMPBuilder.createTeams(Builder, BodyGenCB, nullptr, F->arg_begin()));
```
Comment (kiranchandramohan) [inline llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:4104]: Please add a comment for what is `nullptr` and why using `F->arg_begin()` is fine. Same for the `createTeams` invocation in the other tests.

**Review by kiranchandramohan — COMMENTED** [review]
Is this change for OMP 5.0 and before?

File: llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h (hunk 1 above)
Comment (shraiysh) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1925]: Lowerbound isn't handled for clang either - it is not even recognized as syntactically correct. I don't know what the generated code should look like, as `@__kmpc_push_num_teams` does not accept a lowerbound on number of teams. So, I kept it unimplemented for future work.

Comment (kiranchandramohan) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1925]: So, there are two functions in the OpenMP runtime. It is OK to leave with a TODO, just write a status of the support in Clang and the OpenMP runtime.

```
KMP_EXPORT void __kmpc_push_num_teams(ident_t *loc, kmp_int32 global_tid,
                                      kmp_int32 num_teams,
                                      kmp_int32 num_threads);

/* Function for OpenMP 5.1 num_teams clause */
KMP_EXPORT void __kmpc_push_num_teams_51(ident_t *loc, kmp_int32 global_tid,
                                         kmp_int32 num_teams_lb,
                                         kmp_int32 num_teams_ub,
                                         kmp_int32 num_threads);
```

Comment (shraiysh) [reply llvm/include/llvm/Frontend/OpenMP/OMPIRBuilder.h:1925]: Oh, I had not seen this. I think I will update the PR to handle this. Thank you for pointing this out! 👍

**Review by kiranchandramohan — APPROVED** [review]
This is OK with me.

The only question is whether we should use the `__kmpc_push_num_teams` in the case when the `lowerbound` is not specified.

Please wait for a day to see whether there is any opinion on this topic.

**Pull request merged**